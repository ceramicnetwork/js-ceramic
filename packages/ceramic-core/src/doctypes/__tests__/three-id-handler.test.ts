import ThreeIdHandler from '../three-id-handler'
import CID from 'cids'

jest.mock('../../user')
import User from '../../user'
import DoctypeHandler from "../doctype-handler"
jest.mock('did-jwt', () => ({
  // TODO - We should test for when this function throws as well
  verifyJWT: (): any => 'verified'
}))

const FAKE_CID_1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsr')
const FAKE_CID_3 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlss')
const RECORDS = {
  genesis: { doctype: '3id', owners: [ '0x123' ], content: { publicKeys: { test: '0xabc' } } },
  r1: {
    desiredContent: { publicKeys: { test: '0xabc' }, other: 'data' },
    record: { content: [ { op: 'add', path: '/other', value: 'data' } ], id: FAKE_CID_1, prev: FAKE_CID_1, header: 'aaaa', signature: 'cccc' }
  },
  r2: { record: {}, proof: { blockNumber: 123456 } }
}

describe('ThreeIdHandler', () => {
  let user: User, handler: DoctypeHandler

  beforeAll(() => {
    user = new User()
    user.sign = jest.fn(async () => {
      // fake jwt
      return 'aaaa.bbbb.cccc'
    })
  })

  beforeEach(() => {
    handler = new ThreeIdHandler()
  })

  it('is constructed correctly', async () => {
    expect(handler.doctype).toEqual('3id')
  })

  it('user is set correctly', async () => {
    handler.user = user
    expect(handler.user).toEqual(user)
  })

  it('makes genesis record correctly', async () => {
    await expect(handler.makeGenesis(RECORDS.genesis.content)).rejects.toThrow(/The owner/)
    const record = await handler.makeGenesis(RECORDS.genesis.content, RECORDS.genesis.owners)
    expect(record).toEqual(RECORDS.genesis)
  })

  it('applies genesis record correctly', async () => {
    const state = await handler.applyGenesis(RECORDS.genesis, FAKE_CID_1)
    expect(state).toMatchSnapshot()
  })

  it('makes signed record correctly', async () => {
    const state = await handler.applyGenesis(RECORDS.genesis, FAKE_CID_1)
    await expect(handler.makeRecord(state, RECORDS.r1.desiredContent)).rejects.toThrow(/No user/)
    handler.user = user
    const record = await handler.makeRecord(state, RECORDS.r1.desiredContent)
    expect(record).toEqual(RECORDS.r1.record)
  })

  it('applies signed record correctly', async () => {
    let state = await handler.applyGenesis(RECORDS.genesis, FAKE_CID_1)
    state = await handler.applySigned(RECORDS.r1.record, FAKE_CID_2, state)
    expect(state).toMatchSnapshot()
  })

  it('applies anchor record correctly', async () => {
    let state = await handler.applyGenesis(RECORDS.genesis, FAKE_CID_1)
    state = await handler.applySigned(RECORDS.r1.record, FAKE_CID_2, state)
    state = await handler.applyAnchor(RECORDS.r2.record, RECORDS.r2.proof, FAKE_CID_3, state)
    expect(state).toMatchSnapshot()
  })
})
