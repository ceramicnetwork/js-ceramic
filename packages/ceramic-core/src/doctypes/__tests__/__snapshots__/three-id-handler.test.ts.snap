// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeIdHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
  },
  "anchorStatus": 3,
  "content": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "3id",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          148,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          148,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextOwners": undefined,
  "owners": Array [
    "0x123",
  ],
  "signature": 2,
}
`;

exports[`ThreeIdHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "3id",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextContent": null,
  "owners": Array [
    "0x123",
  ],
  "signature": 0,
}
`;

exports[`ThreeIdHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "3id",
  "log": Array [
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
    Object {
      "codec": "dag-pb",
      "hash": Object {
        "data": Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          148,
        ],
        "type": "Buffer",
      },
      "version": 1,
    },
  ],
  "nextContent": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "nextOwners": undefined,
  "owners": Array [
    "0x123",
  ],
  "signature": 2,
}
`;
